File: heap/lv2_더맵게.py
# https://programmers.co.kr/learn/courses/30/lessons/42626
import heapq

def solution(scoville, K):
    heapq.heapify(scoville)
    count = 0
    while True:
        first = heapq.heappop(scoville)
        if len(scoville) == 0 and first < K:
            return -1
        elif first >= K:
            return count
        else:
            second = heapq.heappop(scoville)
            new = first + 2*second
            heapq.heappush(scoville, new)
            count += 1

File: heap/lv2_더맵게.java
// https://programmers.co.kr/learn/courses/30/lessons/42626
import java.util.PriorityQueue;


class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>();
        for (int scov: scoville) {
            priorityQueue.add(scov);
        }
        while (priorityQueue.peek() < K) {
            if (priorityQueue.size() == 1) {
                if (priorityQueue.poll() < K) {
                    return -1;
                } else {
                    return answer;
                }
            }
            int first = priorityQueue.poll();
            int second = priorityQueue.poll();
            int new_ = first + (2 * second);
            priorityQueue.add(new_);
            answer ++;
        }
        return answer;
    }
}
